### ISSUE TITLE
EG high memory usage

### ISSUE THREAD

**Issue by qicz** [envoyproxy/gateway#1613]
*Version*

the main branch

*Resources Settings*

limits: memory 1Gi,  CPU 500m

*Platform*

Openshift

*Description*:
The current EG implementation watches a lot of resources, such as HTTPRoute, GPCRoute, EnvoyProxy, etc., and a copy of these resources will be reserved in the EG memory, causing the EG memory usage to be very high. When the user's HTTPRoute and other resources continue to increase, the memory of EG will take up more, and the user needs to adjust the resource limits dynamically ?

![image](https://github.com/envoyproxy/gateway/assets/2174082/34d964e8-23b9-4945-8e5c-bf4872747ffe)

When any resource changes, EG will trigger the behavior of these changes, such as translation and infrastructure reconstruction. 

https://github.com/envoyproxy/gateway/blob/6af516d9b36b237c7da44a087eafc57bf3d7d9a2/internal/provider/kubernetes/controller.go#L287-L290

https://github.com/envoyproxy/gateway/blob/6af516d9b36b237c7da44a087eafc57bf3d7d9a2/internal/gatewayapi/runner/runner.go#L51-L66

These are actually unnecessary. Is it possible to separate watch for related resources or separate store and subscribe when resources really change?

like this:

```go
if val, ok := r.resources.GatewayAPIHTTPRoutesResources.Load(acceptedGC.Name); !ok || cmp.Diff(val, httpRoutes) != "" {
   r.resources.GatewayAPIHTTPRoutesResources.Store(acceptedGC.Name, httpRoutes)
}
```

**Comment by qicz** [comment]
> thanks for raising this issue @qicz . Can you also highlight
> 
> 1. Number of resources you have running - Gateways, HTTPRoutes, etc ?
> 2. Version of EG you are running
> 
> Why is the memory increasing when the number of resources are increasing ?
> 
> * As you highlighted above, EG store relevant resources in the `watchable` store, so increase in resources increases the memory usage of the EG.
> * This is an intentional design decision to
>   
>   * Reduce CPU - Caching helps us avoid translations if input resources haven't changed
>   * Reduce I/O - Caching helps us avoid writes to Envoy (xds) or K8s API Server (status) if output resources haven't changed
>     which comes at a cost of memory usage.
> 
> How can we deal with this memory usage at high scale ?
> 
> * Lets set the default memory limits to a higher value https://github.com/envoyproxy/gateway/blob/a5982c395cb5845f53b7c25fec0368222bf33ed3/charts/gateway-helm/values.tmpl.yaml#L8
>    128M is too low imo
> * Reduce Controller Runtime Cache - limit EG to a subset of namespaces [Implement Watch Namespaces #1610](https://github.com/envoyproxy/gateway/pull/1610). This won't affect the `watchable` store since its only saving the
>   relevant/accepted resources
> * @chauhanshubham are they any further optimizations we can make in the provider / reconcile logic ?
> 
> the change you've highlighted above with `cmp.Diff()` is inherently handled by `watchable` which is one of the primary reasons we chose to use it in this project. https://github.com/envoyproxy/gateway/blob/main/docs/latest/design/watching.md

1. just 69 HTTPRoute
<img width="1173" alt="image" src="https://github.com/envoyproxy/gateway/assets/2174082/96b78c07-85cb-4013-93e8-b55d9e8fe723">
2. EG version: the main branch
3. resources limits settings:
<img width="1421" alt="image" src="https://github.com/envoyproxy/gateway/assets/2174082/f49767dc-011d-40ac-bfeb-8b231b0c98f0">


IMO #1617 can not resolve this thoroughly.

**Comment by qicz** [comment]
fix the sa create or update bug like #1609, the limits just need `1Gi`, but the EG occupies the same high mem.

**Comment by arkodg** [comment]
@qicz would be great if you can attach the `pprof` you shared earlier which highlights the high mem usage due to the secrets being cached. We can then see if
* we are caching relevant secrets or all secrets https://github.com/envoyproxy/gateway/blob/708287e53af5146873c91c9a4cda2a9ed8126f03/internal/provider/kubernetes/predicates.go#L71
* if above logic is in fact right, we can attempt to skip caching `secrets` and `configmaps` and directly retrieve them from API server instead of using the `Cache`

**Comment by arkodg** [comment]
has this issue been resolved @qicz ?

**Comment by qicz** [comment]
> has this issue been resolved @qicz ?

fixed by #1594 and #1609 , and I will continue to follow EG performance issues
